modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1245 @@
+  inline proc acquireIfNonEmpty(newStatus) {
+    while !isEmpty {
+      if currentStatus == STATUS_UNLOCKED && acquireWithStatus(newStatus) {
+        if isEmpty {
+          releaseStatus();
+          return false;
+        } else {
+          return true;
+        }
+      }
+
+      chpl_task_yield();
+    }
+
+    return false;
+  }
+
+  inline proc isUnlocked {
+    return status.read() == STATUS_UNLOCKED;
+  }
+
+  inline proc currentStatus {
+    return status.read();
+  }
+
+  inline proc releaseStatus() {
+    status.write(STATUS_UNLOCKED);
+  }
+
+  inline proc transferElements(destPtr, n, locId = here.id) {
+    var destOffset = 0;
+    var srcOffset = 0;
+    while destOffset < n {
+      if headBlock == nil || isEmpty {
+        halt(here, ": Attempted transfer ", n, " elements to ", locId, " but failed... destOffset=", destOffset);
+      }
+
+      var len = headBlock.size;
+      var need = n - destOffset;
+      // If the amount in this block is greater than what is left to transfer, we
+      // cannot begin transferring at the beginning, so we set our offset from the end.
+      if len > need {
+        srcOffset = len - need;
+        headBlock.size = srcOffset;
+        __primitive("chpl_comm_array_put", headBlock.elems[srcOffset], locId, destPtr[destOffset], need);
+        destOffset = destOffset + need;
+      } else {
+        srcOffset = 0;
+        headBlock.size = 0;
+        __primitive("chpl_comm_array_put", headBlock.elems[srcOffset], locId, destPtr[destOffset], len);
+        destOffset = destOffset + len;
+      }
+
+      // Fix list if we consumed last one...
+      if headBlock.isEmpty {
+        var tmp = headBlock;
+        headBlock = headBlock.next;
+        delete tmp;
+
+        if headBlock == nil then tailBlock = nil;
+      }
+    }
+
+    nElems.sub(n : uint);
+  }
+
+  proc addElementsPtr(ptr, n, locId = here.id) {
+    var offset = 0;
+    while offset < n {
+      var block = tailBlock;
+      // Empty? Create a new one of initial size
+      if block == nil then {
+        tailBlock = new BagSegmentBlock(eltType, INITIAL_BLOCK_SIZE);
+        headBlock = tailBlock;
+        block = tailBlock;
+      }
+
+      // Full? Create a new one double the previous size
+      if block.isFull {
+        block.next = new BagSegmentBlock(eltType, block.cap * 2);
+        tailBlock = block.next;
+        block = block.next;
+      }
+
+      var nLeft = n - offset;
+      var nSpace = block.cap - block.size;
+      var nFill = min(nLeft, nSpace);
+      __primitive("chpl_comm_array_get", block.elems[block.size], locId, ptr[offset], nFill);
+      block.size = block.size + nFill;
+      offset = offset + nFill;
+    }
+
+    nElems.add(n : uint);
+  }
+
+  inline proc takeElements(n) {
+    var iterations = n : int;
+    var arr : [{0..#n : int}] eltType;
+    var arrIdx = 0;
+
+    for 1 .. n : int{
+      if isEmpty {
+        halt("Attempted to take ", n, " elements when insufficient");

```

Comment:
Similarly this and other halt messages should at least be preceded with something like "Internal DistributedBag Error:"